Assets/BossRoom/Scripts/Shared/ApplicationController.cs

```diff
@@ -0,0 +1,136 @@
+
+        [SerializeField] private GameObject[] m_AutoInjected;
+
+        private void Awake()
+        {
+            Application.wantsToQuit += OnWantToQuit;
+
+            DontDestroyOnLoad(gameObject);
+            DontDestroyOnLoad(m_UpdateRunner.gameObject);
+
+            var scope = DIScope.RootScope;
+
+            scope.BindInstanceAsSingle(this);
+            scope.BindInstanceAsSingle(m_UpdateRunner);
+            scope.BindInstanceAsSingle(m_GameNetPortal);
+            scope.BindInstanceAsSingle(m_ClientNetPortal);
+
+            //the following singletons represent the local representations of the lobby that we're in and the user that we are
+            //they can persist longer than the lifetime of the UI in MainMenu where we set up the lobby that we create or join
+            scope.BindAsSingle<LobbyUser>();
+            scope.BindAsSingle<LocalLobby>();
+
+            //this message channel is essential and persists for the lifetime of the lobby and relay services
+            scope.BindMessageChannel<UnityServiceErrorMessage>();
+
+            //all the lobby service stuff, bound here so that it persists through scene loads
+            scope.BindAsSingle<AuthenticationAPIInterface>(); //a manager entity that allows us to do anonymous authentication with unity services
+            scope.BindAsSingle<LobbyServiceData>();
+            scope.BindAsSingle<LobbyContentHeartbeat>();
+            scope.BindAsSingle<LobbyAPIInterface>();
+            scope.BindAsSingle<LobbyAsyncRequests>();
+
+
+            scope.FinalizeScopeConstruction();
+
+            foreach (var o in m_AutoInjected)
+            {
+                scope.InjectIn(o);
+            }
+
+            m_LocalLobby = scope.Resolve<LocalLobby>();
+            m_LobbyAsyncRequests = scope.Resolve<LobbyAsyncRequests>();
+            m_LobbyContentHeartbeat = scope.Resolve<LobbyContentHeartbeat>();
+            m_LocalUser = scope.Resolve<LobbyUser>();
+        }
+
+        private void Start()
+        {
+            SceneManager.LoadScene("MainMenu");
+        }
+
+        private void OnDestroy()
+        {
+            ForceLeaveLobbyAttempt();
+            DIScope.RootScope.Dispose();
+        }
+
+        /// <summary>
+        ///     In builds, if we are in a lobby and try to send a Leave request on application quit, it won't go through if we're quitting on the same frame.
+        ///     So, we need to delay just briefly to let the request happen (though we don't need to wait for the result).
+        /// </summary>
+        private IEnumerator LeaveBeforeQuit()
+        {
+            ForceLeaveLobbyAttempt();
+            yield return null;
+            Application.Quit();
+        }
+
+        private bool OnWantToQuit()
+        {
+            var canQuit = string.IsNullOrEmpty(m_LocalLobby?.LobbyID);
+            StartCoroutine(LeaveBeforeQuit());
+            return canQuit;
+        }
+
+        private void ForceLeaveLobbyAttempt()
```

Comment:
Should this really be in ApplicationController?

Reply:
Triggered from here - yes, that logic specifically - it should be in some of the Lobby-related classes